gap.go

```diff
@@ -268,26 +305,43 @@ func (buf *rawAdvertisementPayload) HasServiceUUID(uuid UUID) bool {
 }
 
 // ManufacturerData returns the manufacturer data in the advertisement payload.
-func (buf *rawAdvertisementPayload) ManufacturerData() []ManufacturerDataElement {
-	var manufacturerData []ManufacturerDataElement
-	for index := 0; index < int(buf.len)+4; index += int(buf.data[index]) + 1 {
-		fieldLength := int(buf.data[index+0])
-		if fieldLength < 3 {
-			continue
-		}
-		fieldType := buf.data[index+1]
-		if fieldType != 0xff {
-			continue
-		}
-		key := uint16(buf.data[index+2]) | uint16(buf.data[index+3])<<8
-		manufacturerData = append(manufacturerData, ManufacturerDataElement{
-			CompanyID: key,
-			Data:      buf.data[index+4 : index+fieldLength+1],
+func (buf *rawAdvertisementPayload) ManufacturerData() []AdvertismentDataElement {
+	var manufacturerData []AdvertismentDataElement
+	for _, data := range buf.findAllFields(0xFF) { 
+		manufacturerData = append(manufacturerData, AdvertismentDataElement{
+			UUID: New16BitUUID(uint16(data[0])+(uint16(data[1])<<8)),
+			Data: data[2:],
```

Comment:
The old code has been written in a way that avoids heap allocations. The rewrite introduces new heap allocations, which I think is a regression.

I agree the code is somewhat hard to read. It might be possible to make it easier to read using function iterators, though I'm not entirely sure about that.